Source/WebCore/style/calc/StyleCalculationTree.cpp

```diff
@@ -213,6 +189,413 @@ TextStream& operator<<(TextStream& ts, const Tree& tree)
+
+    auto combinedChildrenSize = root.children.size() - childrenToRemoveTotal;
+
+    // If the new size is 0, we removed too much. Return a single 0 value of type `dimension` to keep things valid. A value of type `dimension` is returned because the only kind of node that can be removed is of type `dimension`.
+    if (!combinedChildrenSize)
+        return { makeChild(Dimension { .value = 0 }) };
+
+    // If the new size is 1, we know there is one child, we just don't know which one yet.
+    if (combinedChildrenSize == 1) {
+        for (size_t i = 0; i < root.children.size(); ++i) {
+            auto replacement = WTF::switchOn(root.children[i],
+                [&]<Numeric T>(const T& child) -> std::optional<Child> {
+                    auto& firstInstance = firstInstances[static_cast<uint8_t>(toNumericIdentity(child))];
+                    ASSERT(firstInstance.offset);
+
+                    // If the stored offset for this type is set to this index and it's not one that can be removed, this is the 1 child to return.
+                    if ((firstInstance.offset - 1) == i && !firstInstance.canRemove)
+                        return { WTFMove(root.children[i]) };
+
+                    // Otherwise, it's one that can be dropped.
+                    return { };
+                },
+                [&](const auto&) -> std::optional<Child> {
+                    return { WTFMove(root.children[i]) };
+                }
+            );
+            if (replacement)
+                return { WTFMove(*replacement) };
+        }
+    }
+
+    Vector<Child> combinedChildren;
+    combinedChildren.reserveInitialCapacity(combinedChildrenSize);
+
+    for (size_t i = 0; i < root.children.size(); ++i) {
+        WTF::switchOn(root.children[i],
+            [&]<Numeric T>(const T& child) {
+                auto& firstInstance = firstInstances[static_cast<uint8_t>(toNumericIdentity(child))];
+                ASSERT(firstInstance.offset);
+
+                // If the stored offset for this type is set to this index and it's not one that can be removed, append the child as normal
+                if ((firstInstance.offset - 1) == i && !firstInstance.canRemove) {
+                    combinedChildren.append(WTFMove(root.children[i]));
+                    return;
+                }
+
+                // Otherwise, it's one that can be dropped.
+            },
+            [&](const auto&) {
+                combinedChildren.append(WTFMove(root.children[i]));
+            }
+        );
+    }
+    root.children = WTFMove(combinedChildren);
+
+    return { };
+}
+
+static std::optional<Child> simplify(Negate& root)
+{
+    // This is almost identical to the Negate simplification in CSSCalcTree+Simplification.cpp with the following differences:
+    // - It operates on Style::Calculation::Negate, not CSSCalc::Negate.
+    // - It returns a Style::Calculation::Child, not CSSCalc::Child.
+    // - It does not take SimplificationOptions.
```

Comment:
The comment could include some example simplifications to make it less abstract.